hazelcast/src/main/java/com/hazelcast/sql/impl/exec/fetch/Fetch.java

```diff
@@ -0,0 +1,133 @@
+
+package com.hazelcast.sql.impl.exec.fetch;
+
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.row.EmptyRowBatch;
+import com.hazelcast.sql.impl.row.ListRowBatch;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.row.RowBatch;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * LIMIT/OFFSET core support.
+ */
+public class Fetch {
+
+    private final Expression<?> fetch;
+    private final Expression<?> offset;
+
+    private long fetchValue;
+    private long offsetValue;
+
+    private long offsetApplied;
+    private long fetchApplied;
+
+    public Fetch(Expression<?> fetch, Expression<?> offset) {
+        this.fetch = fetch;
+        this.offset = offset;
+    }
+
+    public void setup(ExpressionEvalContext context) {
+        fetchValue = getFetchValue(context, fetch);
+        offsetValue = getOffsetValue(context, offset);
+    }
+
+    public static long getFetchValue(ExpressionEvalContext context, Expression<?> fetch) {
+        return fetch == null ? Long.MAX_VALUE : eval(fetch,  context);
+    }
+
+    public static long getOffsetValue(ExpressionEvalContext context, Expression<?> offset) {
+        return offset == null ? 0 : eval(offset, context);
+    }
+
+    public RowBatch apply(RowBatch batch) {
+        if (fetchApplied == fetchValue) {
+            return EmptyRowBatch.INSTANCE;
+        }
+
+        long toOffset = offsetValue - offsetApplied;
+
+        if (toOffset >= batch.getRowCount()) {
+            // Skip the whole batch
+            offsetApplied += batch.getRowCount();
+
+            return EmptyRowBatch.INSTANCE;
+        }
+
+        // Mark offset as finished and adjust the start index accordingly.
+        int startIndex;
+
+        if (toOffset != 0L) {
+            offsetApplied = offsetValue;
+
+            // toOffset is smaller than the batch size and casting to int is safe
+            startIndex = (int) toOffset;
+        } else {
+            startIndex = 0;
+        }
+
+        // Calculate the end position.
+        long toFetch = fetchValue - fetchApplied;
+        long canFetch = batch.getRowCount() - startIndex;
+
+        if (toFetch >= canFetch) {
+            fetchApplied += canFetch;
+
+            // Return the whole remainder.
+            if (startIndex == 0) {
+                return batch;
+            } else {
+                return slice(batch, startIndex, (int) canFetch);
+            }
+        } else {
+            fetchApplied += toFetch;
+
+            return slice(batch, startIndex, (int) toFetch);
+        }
+    }
+
+    public boolean isDone() {
+        return fetchApplied == fetchValue;
+    }
+
+    private static long eval(Expression<?> expression, ExpressionEvalContext context) {
+        assert expression != null;
+
+        Object val = expression.eval(FetchRow.INSTANCE, context);
+        assert val != null && val instanceof Number;
```

Comment:
How do we know that `val` is not null here? Imagine, that we have the following query 
```
SELECT ...
LIMIT (1 + ?) ...
```
If the parameter happened to be `null`, the whole expression will evaluate to `null`.

Reply:
Now I see that we do it in the validator. Resolving.